### ISSUE TITLE
Memory leak in OpenWrt when using Unbound for DNS over TLS

### ISSUE THREAD

**Issue by EricLuehrsen** [openssl/openssl#7128]
The issue is in the discovery phase right now, but looking for insight. OpenWrt users are reporting system crashes due to running out of RAM when they enable TLS over DNS in Unbound. Default builds are with OpenSSL 1.0.2p. OpenWrt is attempting to move to OpenSSL 1.1.0, but it is still held up due to the multitude of packages to port. "lib(7):func(100):reason(65)" can be decoded to buffer resizing in "openssl/crypto/buffer/buffer.c: int BUF_MEM_grow(BUF_MEM *str, size_t len);"
```
Sat Sep  1 17:47:29 2018 daemon.notice unbound: [24512:0] notice: ssl handshake failed 1.0.0.1 port 853
Sat Sep  1 17:47:29 2018 daemon.err unbound: [24512:0] error: ssl handshake failed crypto error:07064041:lib(7):func(100):reason(65)
```
see: https://www.nlnetlabs.nl/bugs-script/show_bug.cgi?id=4149
see: https://github.com/openwrt/packages/issues/6925
see: https://github.com/openwrt/packages/issues/6957

**Comment by mattcaswell** [comment]
So, I agree this does sound like a memory leak somewhere. The problem is where (the failure in OpenSSL's `BUF_mem_grow` does not necessarily point to it being an OpenSSL leak). Presumably this is new behaviour - are you able to narrow it down to a specific OpenSSL version or unbound change? If the problem goes away by using an older OpenSSL version (for example) then that might add weight to the idea that it is an OpenSSL leak introduced by a recent change in OpenSSL. OTOH if the issue remains while running with an older OpenSSL version then that might tend to indicate an unbound issue.

Either way - have you tried running under valgrind or similar? That might help indicate the source of the original leaking allocations.

**Comment by EricLuehrsen** [comment]
Seems valgrind doesn't like on target programs compiled with `-Os` necessary to fit. It wasn't able to get past config file load before getting confused by my guess. I realize the dwarf won't be there either, but it would have been nice to gets hits or not. So its the long way of rebuilding the router with reverted OpenSSL commits.

**Comment by EricLuehrsen** [comment]
1.0.2o repeats the error

**Comment by EricLuehrsen** [comment]
OpenSSL 1.1.0i repeats the error so it looks like Unbound unless it is an API used as documented but documentation doesn't meet intent.

**Comment by EricLuehrsen** [comment]
The witch hunt is over and it was Unbound. TLS state was not entirely mopped up during TCP connection time outs. Not a pure leak and memory was referenced, but logical use was stale. Resolved to openwrt/packages#7112 and also in Unbound trunk.

**Comment by noloader** [comment]
> Seems valgrind doesn't like on target programs compiled with -Os necessary to fit

When using Valgrind you should compile at `-O1`. Also see the Valgrind [Valgrind Quick Start Guide](https://valgrind.org/docs/manual/quick-start.html). Otherwise result accuracy suffers.
